airflow/contrib/operators/gcp_transfer_operator.py

```diff
@@ -0,0 +1,856 @@
+    :param gcs_bucket: The destination Google Cloud Storage bucket
+        where you want to store the files. (templated)
+    :type gcs_bucket: str
+    :param project_id: Optional ID of the Google Cloud Platform Console project that
+        owns the job
+    :type project_id: str
+    :param aws_conn_id: The source S3 connection
+    :type aws_conn_id: str
+    :param gcp_conn_id: The destination connection ID to use
+        when connecting to Google Cloud Storage.
+    :type gcp_conn_id: str
+    :param delegate_to: The account to impersonate, if any.
+        For this to work, the service account making the request must have
+        domain-wide delegation enabled.
+    :type delegate_to: str
+    :param description: Optional transfer service job description
+    :type description: str
+    :param schedule: Optional transfer service schedule;
+        If not set, run transfer job once as soon as the operator runs
+        The format is described
+        https://cloud.google.com/storage-transfer/docs/reference/rest/v1/transferJobs.
+        With two additional improvements:
+
+        * dates they can be passed as :class:`datetime.date`
+        * times they can be passed as :class:`datetime.time`
+
+    :type schedule: dict
+    :param object_conditions: Optional transfer service object conditions; see
+        https://cloud.google.com/storage-transfer/docs/reference/rest/v1/TransferSpec
+    :type object_conditions: dict
+    :param transfer_options: Optional transfer service transfer options; see
+        https://cloud.google.com/storage-transfer/docs/reference/rest/v1/TransferSpec
+    :type transfer_options: dict
+    :param wait: Wait for transfer to finish
+    :type wait: bool
+    """
+
+    template_fields = ('gcp_conn_id', 's3_bucket', 'gcs_bucket', 'description',
+                       'object_conditions')
+    ui_color = '#e09411'
+
+    @apply_defaults
+    def __init__(self,
+                 s3_bucket,
+                 gcs_bucket,
+                 project_id=None,
+                 aws_conn_id='aws_default',
+                 gcp_conn_id='google_cloud_default',
+                 delegate_to=None,
+                 description=None,
+                 schedule=None,
+                 object_conditions=None,
+                 transfer_options=None,
+                 wait=True,
+                 *args,
+                 **kwargs):
+
+        super(S3ToGoogleCloudStorageTransferOperator, self).__init__(
+            *args,
+            **kwargs)
+        self.s3_bucket = s3_bucket
+        self.gcs_bucket = gcs_bucket
+        self.project_id = project_id
+        self.aws_conn_id = aws_conn_id
+        self.gcp_conn_id = gcp_conn_id
+        self.delegate_to = delegate_to
+        self.description = description
+        self.schedule = schedule
+        self.object_conditions = object_conditions
+        self.transfer_options = transfer_options
+        self.wait = wait
+
+    def execute(self, context):
```

Comment:
How about extracting the functionality inside `execute()` to a few methods here?
The `execute()` seems a bit bloated and violates the SRP principle a bit.
I'd suggest extracting the body creation and the later validations to two separate methods.

Reply:
I created a `_create_body` method.